Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs

```diff
@@ -0,0 +1,144 @@
+        {
+            int pipsAdded = 0;
+            string fileName = filePath.ToString(context.PathTable);
+            using (FileStream stream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read))
+            using (RemapReader reader = new RemapReader(pipFragmentContext, stream, context))
+            {
+                while (reader.ReadBoolean())
+                {
+                    pipsAdded++;
+                    var pip = Pip.Deserialize(reader);
+                    pip.ResetPipIdForTesting();
+
+                    bool added = false;
+                    switch (pip.PipType)
+                    {
+                        case PipType.Module:
+                            var modulePip = pip as ModulePip;
+                            added = pipGraphBuilder.AddModule(modulePip);
+                            break;
+                        case PipType.SpecFile:
+                            var specFilePip = pip as SpecFilePip;
+                            added = pipGraphBuilder.AddSpecFile(specFilePip);
+                            break;
+                        case PipType.Value:
+                            var valuePIp = pip as ValuePip;
+                            added = pipGraphBuilder.AddOutputValue(valuePIp);
+                            break;
+                        case PipType.Process:
+                            var p = pip as Process;
+                            added = pipGraphBuilder.AddProcess(p, default);
+                            break;
+                        case PipType.CopyFile:
+                            var copyFile = pip as CopyFile;
+                            added = pipGraphBuilder.AddCopyFile(copyFile, default);
+                            break;
+                        case PipType.WriteFile:
+                            var writeFile = pip as WriteFile;
+                            added = pipGraphBuilder.AddWriteFile(writeFile, default);
+                            break;
+                        case PipType.SealDirectory:
+                            var sealDirectory = pip as SealDirectory;
+                            if (sealDirectory.Kind == SealDirectoryKind.Opaque || sealDirectory.Kind == SealDirectoryKind.SharedOpaque)
+                            {
+                                continue;
+                            }
+
+                            added = true;
+                            var oldDirectory = sealDirectory.Directory;
+                            sealDirectory.ResetDirectoryArtifact();
+                            var mappedDirectory = pipGraphBuilder.AddSealDirectory(sealDirectory, default);
+                            reader.Context.AddMapping(oldDirectory, mappedDirectory);
+                            break;
+                    }
+
+                    if (!added)
+                    {
+                        Logger.Log.FailedToAddFragmentPipToGraph(loggingContext);

```

Comment:
>FailedToAddFragmentPipToGraph [](start = 35, length = 29)

This should include details about the fragment (i.e. path and name) and failing pip.

Reply:
done.

---
In reply to: [289095947](https://github.com/ljcollins25/BuildXL/pull/1#discussion_r289095947) [](ancestors = 289095947)